### PR TITLE
docs: create UPGRADING.md

### PR DESCRIPTION
**Pull request by KwilLuke**
An initial draft of the `UPGRADING.md` file to assist users with upgrading between versions.

I drew inspiration from CometBFT's [`UPGRADING.md` file](https://github.com/cometbft/cometbft/blob/main/UPGRADING.md).

Content for this section drew from [this discussion](https://github.com/kwilteam/kwil-db/discussions/668) and [wiki](https://github.com/kwilteam/kwil-db/wiki/Kwil-v0.8).

The links should work once [docs#105](https://github.com/kwilteam/docs/pull/105) and [docs#108](https://github.com/kwilteam/docs/pull/108) are merged and published.

This is certainly a first draft. Let me know if any information is missing or if you have alternative recommendations for how to present this information.

### CODE REVIEW

File: UPGRADING.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+- The gRPC server (`grpc_listen_addr`) is no longer exposed.
+
+### Node Configuration Changes
+
+In the node's `config.toml` file, the following changes have been made:
+
+- Added `app.rpc_timeout`, which imposes a timeout on RPC requests. The default is 45 seconds.
+- Added `app.db_read_timeout`, which imposes a timout on read-only DB transactions (i.e. actions/procedures with a [`view` tag](https://docs.kwil.com/docs/kuneiform/supported-features#access-tags) or an ad-hoc SELECT query). The default 5 seconds.
+- Added `rpc.broadcast_tx_timout`, which imposes a timeout for awaiting transaction confirmation when using the `--sync` flag in `kwil-cli`. The default is 10 seconds.
+- Added `app.snapshots` to support snapshot creation. See the `StateSyncConfig` struct for more information.
```
Comment (charithabandi) [inline UPGRADING.md:61]: app.snapshots -> enables creation of snapshots --> look for `SnapshotConfig`
app.statesync --> used to join a network by requesting peers for snapshots and syncs with the network.  ---> look for `StateSyncConfig`

Probably mention these 2 seperately.

File: UPGRADING.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+- Added `app.snapshot_file` to suport starting a network with a state at genesis. This is to support the Network Migrations feature.
```
Comment (charithabandi) [inline UPGRADING.md:62]: should we rename it to genesis_snapshot_file or something like that to remove confusion with the other snapshots?

File: UPGRADING.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+### CLI Changes
+
+- The `kwil-cli` has been updated to use the JSON-RPC server.
+- The `--kwil-provider` flag is DEPRECATED and will be removed in the subsequent release. Use the `--provider` flag instead.
+- Added `kwil-cli utils parse` to parse a Kuneiform file and output the schema as JSON.
+- Added `kwil-cli utils decode-tx` to decode a transaction from a base64 string.
```
Comment (charithabandi) [inline UPGRADING.md:69]: kwil-cli utils query-tx would not return tx payload if the tx is still in the mempool and not mined yet. 

This is a behavior change from before.

File: UPGRADING.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+- There are many new types and fields on the `Schema` struct to support procedures and strongly typed values. Changes include:
+    - `Owner` is now `HexBytes` instead of `byte[]`. `HexBytes` is an alias for `byte[]`; however, when marshaled to JSON, it is represented as a hex string.
+    - `Table.Column.Type` now uses the `DataType` struct instead of a `string` to support strongly typed values.
+    - Added the `Procedures` field to the `Schema` struct to support procedures.
+    - Added the `ForeignProcedures` field to the `Schema` struct to support calling procedures that are defined in other schemas.
+
+#### `core/types/transactions`
+
+- **Note**: Most users do not need to be concserned with these changes as the `core/client` and `core/rpc/...` packages handle their use.
```
Comment (charithabandi) [inline UPGRADING.md:118]: concerned*

File: UPGRADING.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+- The `kwil-cli` has been updated to use the JSON-RPC server.
+- The `--kwil-provider` flag is DEPRECATED and will be removed in the subsequent release. Use the `--provider` flag instead.
+- Added `kwil-cli utils parse` to parse a Kuneiform file and output the schema as JSON.
+- Added `kwil-cli utils decode-tx` to decode a transaction from a base64 string.
+
+#### Breaking Changes
+
+- Because Kwil-CLI now communicates over JSON-RPC, you should pass the JSON-RPC listen address to the CLI with the `--provider` flag or `kwil-cli configure` command.
+- For uncomfirmed transactions, `kwild` now returns `null` for the `tx` field. Using older versions of `kwil-cli` to query unconfirmed transactions may have unexpected behavior.
```
Comment (charithabandi) [inline UPGRADING.md:74]: Maybe specify that the query-tx api returns no tx data for unconfirmed transactions.

File: UPGRADING.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+## v0.8.0
+
+Kwil v0.8 introduces substantial changes to the Kuneiform language (SQL smart contracts), an improved network migration process, a system for coordinating consensus rules changes, and a JSON-RPC server.
+
+### Upgrading from v0.7.x
+
+Nodes running `kwild` v0.8 are not compatible with nodes running `kwild` v0.7.
+
+If a v0.7 network uses schemas that **are** compatible with v0.8, follow the [network migrations guide](https://docs.kwil.com/docs/node/network-migrations) to export the network data (CometBFT and Postgres) and import it into a new v0.8 network.
```
Comment (brennanjl) [inline UPGRADING.md:13]: It does not export comet data

Comment (brennanjl) [reply UPGRADING.md:13]: Also, maybe add a note here to see the breaking changes to determine whether a schema is compatible?

Comment (brennanjl) [reply UPGRADING.md:13]: Also, to test whether a schema is compatible, users should, using the `kwil-cli` for v0.8, run `kwil-cli utils parse ./schema.kf`. If it parses with no errors, their schema is compatible.

File: UPGRADING.md [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+- SQL statements that are valid SQL but have invalid predicates will fail during parse, instead of at runtime.
+- Fixed edge cases where users could perform cartesian joins. Now, one side of a join must be a column of unique values.
+- Fixed a handful of cases that could lead to non-deterministic query results:
+    - Disallowing conflicting column names when returned from a query.
+    - Not returning unnamed columns in query results.
+    - Applying deterministic ordering for joins against subqueries.
+
+### Network Migrations
+
+- Kwil has a network migration process that allows for exporting network data (CometBFT and Postgres) and importing it into a new network. This can be used to upgrade a network to a new version (assuming the old version does not use features that are incompatible with the new version) or to create a new network with the same data. - **[Learn more](https://docs.kwil.com/docs/node/network-migrations)**
```
Comment (brennanjl) [inline UPGRADING.md:36]: once again, not comet data

File: UPGRADING.md [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+- The HTTP server (`http_listen_addr`) is DEPRECATED and will be removed in the susbequent release.
+- The gRPC server (`grpc_listen_addr`) is no longer exposed.
+
+### Node Configuration Changes
+
+In the node's `config.toml` file, the following changes have been made:
+
+- Added `app.rpc_timeout`, which imposes a timeout on RPC requests. The default is 45 seconds.
+- Added `app.db_read_timeout`, which imposes a timout on read-only DB transactions (i.e. actions/procedures with a [`view` tag](https://docs.kwil.com/docs/kuneiform/supported-features#access-tags) or an ad-hoc SELECT query). The default 5 seconds.
```
Comment (brennanjl) [inline UPGRADING.md:59]: fyi, this link has been removed from docs. I don't think you need to link here; anybody who is reading our upgrading would know what `view` means

File: UPGRADING.md [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+```go
+type Signature struct {
+    Signature []byte `json:"sig"` // was `json:"signature_bytes"`
+    Type string `json:"type"` // was `json:"signature_type"`
+}
+```
+
+#### `core/types`
+
+- The `Schema` struct and all of it's composing types are now mirrored in `core/types/transactions`.
```
Comment (brennanjl) [inline UPGRADING.md:105]: related, but we used to keep the main schema struct in `common.Schema`. It is now in `core/types.Schema`

File: UPGRADING.md (hunk 5 above)
Comment (KwilLuke) [reply UPGRADING.md:74]: I think that is what: "- For uncomfirmed transactions, `kwild` now returns `null` for the `tx` field. Using older versions of `kwil-cli` to query unconfirmed transactions may have unexpected behavior." Is that not clear?

File: UPGRADING.md (hunk 2 above)
Comment (KwilLuke) [reply UPGRADING.md:62]: I think that could make sense cc @brennanjl.

Comment (jchappelow) [reply UPGRADING.md:64]: Or `genesis_state`

Comment (brennanjl) [reply UPGRADING.md:64]: Sorry did not see this. I think something involving genesis makes more sense here. I actually got it confused with the other snapshot path once

Comment (brennanjl) [reply UPGRADING.md:64]: @KwilLuke can you update the doc here?

**Comment by KwilLuke** [comment]
@brennanjl This should be good to go.

**Review by brennanjl — APPROVED** [review]

**Pull request merged**